### PR TITLE
Add support for properties in channel definitions

### PR DESCRIPTION
**Pull request by cdjackson**
As discussed here -: https://www.eclipse.org/forums/index.php/t/1015656/

I want to do more testing before this is merged - I've added a test case and it seems ok, but I'd also like to do a real-life test in a few days (hopefully!).

Feel free to review...

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Thanks @cdjackson, as I am currently in OH release preparations, I won't be able to review and test this before next week - at a first glance it all looks good through!

**Comment by cdjackson** [comment]
No problem - that’s fine. I might have had a chance to test it more by then :)

File: bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <thing:thing-descriptions bindingId="hue"
 	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
-	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 org.eclipse.smarthome.thing-description.xsd">
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 org.eclipse.smarthome.thing-description-1.0.0.xsd">
```
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml]: Is this change needed?

File: bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelGroupTypeXmlResult.java [hunk 2]
```diff
@@ -11,7 +11,6 @@
 import java.util.List;
```
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelGroupTypeXmlResult.java]: Could you please add the author tag, here?

File: bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/ThingTypesTest.groovy [hunk 3]
```diff
@@ -71,6 +71,13 @@ class ThingTypesTest extends OSGiTest {
             assertThat size(), is(3)
             def colorChannel = it.find { it.id.equals("color") } as ChannelDefinition
             assertThat colorChannel, is(notNullValue())
+           
+            colorChannel.with {
+            	assertThat properties.size(), is(2)
+        		assertThat properties.get("key1"), is("value1")
+        		assertThat properties.get("key2"), is("value2")
```
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/ThingTypesTest.groovy]: I know, we have no common groovy formatter, but here spaces and tabs are mixed. Could you change it to spaces, please.

**Comment by dnobel** [comment]
Hi Chris, 

Thanks for the PR. I´m done with the review and only have minor remarks. Just two conceptional questions:

1) Do we also need properties for channel group definitions?
2) Do you think we should expose the properties via REST as we do it for the thing? As discussed in the forum the properties are mainly there for the internal binding implementation. So I don´t see a need here.

Regards

Dennis

**Comment by kaikreuzer** [comment]
> 2) Do you think we should expose the properties via REST as we do it for the thing?

I do not see a reason to diverge here between the "normal" thing object and its bean representation; so I would vote for adding it there as well.

**Comment by cdjackson** [comment]
Thanks Dennis, I'll take a look at the comments tonight.

1) I personally don't see a need for properties in the channel groups, but of course someone else might in future. I don't mind adding it, or we can leave it until someone asks...

2) I agree with you - I currently see no need to have a REST API as in my mind the properties aren't something the UI needs to know about - it's there for binding configuration... As with anything, that might change, but I'd vote not to have it for now until someone comes up with a use case. I see Kai wants it added though so I'll let you guys decide...

Cheers
Chris

**Comment by dnobel** [comment]
Hi Chris,

I talked to Kai and while discussing the REST stuff we found an inconsistency with the thing properties concept. Currently the thing properties are present at the ThingType and at the Thing. When a thing is created the properties are copied to the thing thing (see ThingFactory).

We think it would be the best to have exactly the same concept for channel properties. So the Channel interface must also be extended with a getProperties() method and in the ThingFactory the properties must be copied from the ChannelDefinition to the Channel.

As we have thing properties in the REST API it would be most convenient to have properties in ChannelTypeBean and ChannelBean, too. Special UIs might know and use binding specific properties.

Do you see a chance to extend the PR as described?

Regards

Dennis

File: bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml (hunk 1 above)
Comment (cdjackson) [inline bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml]: Possibly not, but all the other test thing.xml in this bundle have it specified like this so it seems sensible to make them the same?

**Comment by cdjackson** [comment]
Ok - I think I know what you mean. I'll take a look...

I've update the PR with your previous comments. The only contentious issue is the XML reference - I think I've made this consistent with all other XML files in the bundle which to me makes sense...

**Comment by cdjackson** [comment]
Regarding the REST API.  Does it make sense split the channel properties across both the Thing and the ThingType? It's a simple Map, so I think it makes more sense to only have it in the ThingType interface (ie ChannelTypeBean).

**Comment by cdjackson** [comment]
Actually, there is no ChannelTypeBean - currently channels in ThingTypeBean is a string list. If I change this to make it a ChannelTypeBean, it will be a breaking change for the REST API. Are you happy with this?

The other option is to put it in ThingBean (ie ChannelBean). This would be ok, but it does seem slightly wrong since the properties are part of the ThingType...

**Comment by dnobel** [comment]
> Regarding the REST API. Does it make sense split the channel properties across both the Thing and the ThingType?

It is not only about the REST API. It is about the core API. If the concept of properties is the sam for things and channels it must be present on the meta level (thing type, channel definition) and on the instance level (thing, channel)

> Actually, there is no ChannelTypeBean

But I see a ChannelDefinitionBean, which should have the properties On the Thing level there is the ChannelBean, which could be enriched with properties.

**Comment by cdjackson** [comment]
Hmmm - not sure why it's saying I don't have a CLA...

Anyway - @dnobel - I was wondering if you could take a look at this. It mostly works - the ThingType is ok, and I get the properties in the REST interface. Following a thing create through, it looks like the properties are also being copied across ok in the ThingFactory, but ultimately the properties end up set to null by the time it gets to read back out.
It seems to disappear into the ThingModel and I'm a bit lost in there - can you shed any light on this? Is there anything else (eg in the models) that defines this in any way?

Thanks

**Comment by kaikreuzer** [comment]
> Hmmm - not sure why it's saying I don't have a CLA...

Just ignore this - we know that you have one :-)

File: bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml (hunk 1 above)
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/ConfigDescriptionsTest.bundle/ESH-INF/thing/thing-types.xml]: Yes make sense then.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingFactory.java [hunk 4]
```diff
@@ -193,6 +194,9 @@ private static Channel createChannel(ChannelDefinition channelDefinition, ThingU
             }
         }
 
+        // TODO: Is this safe, or should I make a copy?
+        channelBuilder = channelBuilder.withProperties(channelDefinition.getProperties());
```
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingFactory.java]: Good question. I think a copy would be definitely safer. But I would expect that the channel builder creates a copy.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/Channel.java [hunk 5]
```diff
@@ -49,24 +54,29 @@ public Channel(ChannelUID uid, String acceptedItemType) {
         this.uid = uid;
         this.acceptedItemType = acceptedItemType;
         this.configuration = new Configuration();
+        this.properties = Collections.unmodifiableMap(new HashMap<String, String>(0));
```
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/Channel.java]: The properties of the thing are mutable. So in order to be compliant I would expect the same for the Channel. This is also means, that we need a "setProperty" method similar to the Thing.

**Comment by dnobel** [comment]
Hi Chris, 

besides the mutable properties comment it looks fine for me. I would also expect it to work. I can later try to test it out myself.

Btw: the CLA check failed, because your last commit didn´t had a sign-off. But I guess at the end you need to squash the commits nevertheless.

Regards

Dennis

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/Channel.java (hunk 5 above)
Comment (cdjackson) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/Channel.java]: Ok - I'd considered these properties to be read only, but I'll change this as suggested.

**Comment by cdjackson** [comment]
Thanks Dennis. It's strange - everything looks like it's working through the thing creation, but when I read it out in the REST, it's null (which is strange anyway as I don't see how it should be null - it should be an empty list I think).
The ThingType side is fine - I'm getting the properties there ok.

Cheers
Chris

**Comment by dnobel** [comment]
Hi Chris,

I tested it with your branch and for me it works. I modified the yahoo weather binding and added a property:

```xml
<channel id="pressure" typeId="pressure">
    <properties>
        <property name="a">value1</property>
    </properties>
</channel>
```

I have started the runtime, approved the inbox entry for the yahoo weather discovery result. I debugged it, I have seen that it was copied into the Thing. And if I call rest/things it returns the channels with the properties:

```json
{"type":"NumberItem","name":"yahooweather_weather_20066277_temperature","label":"Temperatur","category":"Temperature","state":"17","tags":[],"stateDescription":{"pattern":"%.1f Â°C","readOnly":true,"options":[]},"groupNames":["yahooweather_weather_20066277"]}],"type":"GroupItem","name":"yahooweather_weather_20066277","label":"Yahoo weather Dortmund, Germany","state":"UNDEF","tags":["thing"],"groupNames":["gGF"]}},{"configuration":{"refresh":60,"location":"12833703"},"properties":{},"statusInfo":{"status":"ONLINE","statusDetail":"NONE"},"UID":"yahooweather:weather:12833703","channels":[{"linkedItems":[],"id":"temperature","itemType":"Number","properties":{}},{"linkedItems":[],"id":"humidity","itemType":"Number","properties":{}},{"linkedItems":[],"id":"pressure","itemType":"Number","properties":{"a":"value1"}}]}]
```

So everything seems to be fine.

Regards

Dennis

**Comment by cdjackson** [comment]
Ok - thanks Dennis. I did basically the same thing, so not so sure why I was seeing null…

Anyway, I’ll give it another try and if it looks good I’ll make the couple of changes you’ve suggest, squash and go...

Thanks.

Cheers
Chris

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/Channel.java (hunk 5 above)
Comment (cdjackson) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/Channel.java]: There are no other setters in this class - also, tags are already 'unmodifiable' - do we really want to add a setter just for the properties?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingFactory.java (hunk 4 above)
Comment (cdjackson) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingFactory.java]: ChannelBuilder doesn't, but Channel constructor does so I think this is ok...

**Comment by cdjackson** [comment]
Dennis. This is good to go I think. I've not added the setter to the Channel class as per my comment above. If you're happy with this then feel free to merge, otherwise we can discuss further...

**Comment by dnobel** [comment]
I don´t have a strong opinion. In order to be compliant to the thing, properties should be editable. But you are right, the whole channel is unmutable at the moment and I don´t see use cases where someone wants to change the channel properties at runtime. 

But we had the same discussion for the thing and at the end we decided to introduce the "setProperty" method. So I tend to introduce the setProperty method on the channel, too. @kaikreuzer What do you think?

**Comment by kaikreuzer** [comment]
I honestly think that the whole "editing thing structure" topic is a bit messy. We started off with immutable objects, because we actually do not want any "user" to expect that he can simply set new values and stuff would be automatically persisted and active - we rather wanted to be in control. So in general, I would treat channels this way as well - so rather make it clear that you have to clone in order to modify it and thus the ChannelBuilder could be rather the tool of choice. Why did we decide for a setProperty on the thing (instead of casting to ThingImpl within the BaseThingHandler.updateProperty method)?

**Comment by dnobel** [comment]
> Why did we decide for a setProperty on the thing (instead of casting to ThingImpl within the BaseThingHandler.updateProperty method)

Because we thought it is very unclean to cast to an internal implementation class. Problem is, that channel, does not even has an interface.

I agree, that the current implementation is a bit messy and confusing. The question is who returns the copy of an object. In enterprise system when using OR mappers like JPA, you always get copies of the entities by the database classes (in our case our registries). So we could discuss that the Registries always returns deep copies of Items, Things etc.

A pattern I have never seen before, but which could also be possible is, that our classes have a clone() method, which returns something like an "MutableThing", that has setters. Then the developer has to clone it when he wants to edit the entity.

We can restart the discussion about the whole concept, but I would suggest to make it independent from this PR. So question remains, should we leave the channel unmutable, or put a setter there.

**Comment by kaikreuzer** [comment]
> So question remains, should we leave the channel unmutable, or put a setter there.

Well, if we currently do not have the need to change properties on channels after they have been created, I'd suggest to go without setters for the moment.

**Comment by cdjackson** [comment]
I would also vote with Kai... While I can see a case where it might be nice to change these properties, my original thought was that they are read-only, and the edit case is really a fringe case. I would therefore be tempted to merge this, and discuss more conceptual changes outside of this PR...

**Comment by dnobel** [comment]
OK, agree.

But one last comment regarding the copying of the properties. You wrote 

> ChannelBuilder doesn't, but Channel constructor does so I think this is ok...

In constructor I can see a `Collections.unmodifiableMap` call, but this does not copy the map. AFAIK It just wraps it in an unmutable map. So I would really expect that the ChannelBuilder or the constructor creates an unmutable deep copy.

**Comment by cdjackson** [comment]
For the copying of properties, I think I'm handling properties in the same way as other elements such as tags. In ThingFactory, tags, and properties, are simply 'got' from somewhere (the thing type or channel definition). We then call ChannelBuilder which in turn creates a Channel - in each of these classes we just pass the reference.

So, I think this should be ok - I think I should treat properties differently as it looks like they are originally instantiated in a similar way, and then treated the same? What do you think?

**Comment by dnobel** [comment]
> For the copying of properties, I think I'm handling properties in the same way as other elements such as tags. In ThingFactory, tags, and properties, are simply 'got' from somewhere (the thing type or channel definition). 

I think it also wrong for there. Better would be to create real copies. But it does not matter, as long as these properties are not modifiable. Because otherwise you would modify the properties of the according ThingType, too.

> So, I think this should be ok - I think I should treat properties differently as it looks like they are originally instantiated in a similar way, and then treated the same? What do you think?

Yes, I think it is OK and the PR can be merged.

**Comment by cdjackson** [comment]
Thanks Dennis - so I don't think there's anything I need to do? The PR is squashed already so I'll leave it to you and Kai to merge.

Cheers
Chris

**Comment by dnobel** [comment]
Just merged. Thanks!
